lms/views/decorators/h_api.py

```diff
@@ -0,0 +1,96 @@
+
+import json
+import requests
+
+from pyramid.httpexceptions import HTTPBadGateway
+from pyramid.httpexceptions import HTTPBadRequest
+from pyramid.httpexceptions import HTTPGatewayTimeout
+
+from lms import util
+from lms.util import MissingToolConsumerIntanceGUIDError
+from lms.util import MissingUserIDError
+
+
+def create_h_user(wrapped):  # noqa: MC0001
+    """
+    Create a user in h if one doesn't exist already.
+
+    Call the h API to create a user for the authorized LTI user, if one doesn't
+    exist already.
+    """
+    def wrapper(request, jwt):  # pylint: disable=too-many-branches
+        params = request.params
+
+        try:
+            oauth_consumer_key = params["oauth_consumer_key"]
+        except KeyError:
+            raise HTTPBadRequest('Required paramer "oauth_consumer_key" missing from LTI params')
+
+        # Only create users for application instances that we've enabled the
+        # auto provisioning features for.
+        enabled_consumer_keys = request.registry.settings["auto_provisioning"]
+        if oauth_consumer_key not in enabled_consumer_keys:
+            return wrapped(request, jwt)
+
+        # Our OAuth 2.0 client_id and client_secret for authenticating to the h API.
+        client_id = request.registry.settings["h_client_id"]
+        client_secret = request.registry.settings["h_client_secret"]
+
+        # The authority that we'll create h users and groups in.
+        authority = request.registry.settings["h_authority"]
+
+        # The URL of h's create user API endpoint.
+        create_user_api_url = request.registry.settings["h_api_url"] + "/users"
+
+        try:
+            username = util.generate_username(params)
+            provider = util.generate_provider(params)
+            provider_unique_id = util.generate_provider_unique_id(params)
+        except MissingToolConsumerIntanceGUIDError:
+            raise HTTPBadRequest('Required parameter "tool_consumer_instance_guid" missing from LTI params')
+        except MissingUserIDError:
+            raise HTTPBadRequest('Required parameter "user_id" missing from LTI params')
+
+        # The user data that we will post to h.
+        user_data = {
+            "username": username,
+            "display_name": util.generate_display_name(params),
+            "authority": authority,
+            "identities": [
+                {
+                    "provider": provider,
+                    "provider_unique_id": provider_unique_id,
+                }
+            ],
+        }
+
+        # Call the h API to create the user in h if it doesn't exist already.
+        try:
+            response = requests.post(
+                create_user_api_url,
+                auth=(client_id, client_secret),
+                data=json.dumps(user_data),
+                timeout=1,
+            )
+        except requests.exceptions.ReadTimeout:
+            raise HTTPGatewayTimeout(explanation="Connecting to Hypothesis failed")
```

Comment:
Do we want this to be a user facing error message?

Reply:
That's what it's meant to be, but I wasn't sure what to say to the user. This message would be shown if the LMS app itself was working fine (and the LTI launch request received was fine and everything) but _h_ seems to be down. I guess it could be preceded by a more generic "Couldn't launch Hypothesis :(" or something?

The various 'Required parameter "..." missing from LTI params' messages above will be seen by users too